governance/program/src/instruction.rs

```diff
@@ -0,0 +1,264 @@
+//! Program instructions
+
+use crate::state::{
+    Vote, MAX_GOVERNANCE_NAME_LENGTH, MAX_PROPOSAL_DESCRIPTION_LINK_LENGTH,
+    MAX_PROPOSAL_INSTRUCTION_DATA_LENGTH, MAX_PROPOSAL_NAME_LENGTH,
+};
+
+/// Instructions supported by the Governance program
+#[derive(Clone)]
+#[allow(clippy::large_enum_variant)]
+pub enum GovernanceInstruction {
+    /// Creates Governance account
+    ///
+    ///   0. `[writable]` Governance account. The account pubkey needs to be set to program-derived address (PDA) with the following seeds:
+    ///           1) 'governance' const prefix,
+    ///           2) Governed Program account key
+    ///   1. `[]` Account of the Program governed by this Governance account
+    ///   2. `[writable]` Program Data account of the Program governed by this Governance account
+    ///   3. `[signer]` Current Upgrade Authority account of the Program governed by this Governance account
+    ///   4. `[]` Governance mint that this Governance uses
+    ///   5. `[signer]` Payer
+    ///   6. `[]` System account
+    ///   7. `[]` Bpf_upgrade_loader account
```

Comment:
This instruction calls` Bpf_upgrade_loader` via CPI to change `upgrade_authority` and CPI `invoke()` needs the called program account

Reply:
Ah I see, I was imagining that the current program upgrade authority would reassign the upgrade authority to the governance *outside* of the governance program.  I'm not a huge fan of reassigning the authority in the `CreateGovernance` instruction.  It's not clear that creating a governance should also change the upgrade authority on a program.

Reply:
In order for `Governance` program to be able to upgrade the governed program we have to transfer its `upgrade_authority` to the `Governance` but the question is when is the best time to do so.

My thinking was that it would be quite elegant that by the fact of creating a Governance for a program you give up your `upgrade_authority` on it at the same time. It also proves that you are program owner and you must sign the transaction.

If you think that's too far fetched and rather happen at later time then at minimum we should check the current upgrade_authority signed the transaction to prevent creating governance for programs you don't own.

Reply:
> If you think that's too far fetched and rather happen at later time then at minimum we should check the current upgrade_authority signed the transaction to prevent creating governance for programs you don't own.

This way is my vote: I first create a program myself, iterate on it, then it gets popular and I want to subject future upgrades of it to governance.  I could do this by assigning the upgrade authority to the Governance program in the same transaction that initializes the Governance account

Reply:
Ok